sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,493 @@
+the smart contract's _identifier_ as input.  A smart contract's
+identifier is a hash of the smart contract's definition, represented
+as a Stacks address with a specific "smart contract" version byte. The
+smart contract identifier is a _principal_.
+
+For example, to call the function `register-name` in a smart contract,
+you would use:
+
+```scheme
+(contract-call
+    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ
+    'register-name
+    name-to-register)
+```
+
+This function returns a boolean-- the return value of the called smart
+contract function. Note that if a called smart contract returns
+`false`, it is guaranteed to not alter any smart contract state
+whatsoever. Of course, any transaction fees paid for the execution
+of that function will not be returned.
+
+The following limitations are imposed on contract calls:
+
+1. No dynamic dispatch. At the time of the smart contract creation,
+   any contracts being called must be specified. Future designs may
+   enable this by allowing contract principals to be supplied as
+   transaction arguments, however, on initial release, we believe
+   dynamic invocation to be too dangerous to support.
+2. Called smart contracts _must_ exist at the time of creation.
+3. No cycles may exist in the call graph of a smart contract. This
+   prevents recursion (and re-entrancy bugs). Such structures can
+   be detected with static analysis of the call graph, and will be
+   rejected by the network.
+
+A key benefit of the static analyzability of this smart contracting
+language is that _all_ transactions that can possibly be called from
+the outermost transaction can be known _a priori_ so that a user can
+be warned about all side effects before signing a transaction.
+
+## Principals and Transaction Verification
+
+The language provides a primitive for checking whether or not the
+smart contract transaction was signed by a particular
+_principal_. Principals are a specific type in the smart contracting
+language which represent a signing entity (roughly equivalent to a
+Stacks address). The signature itself is not checked by the smart
+contract, but by the VM. A smart contract function can use a globally
+defined variable to obtain the current transaction's signer:
+
+```scheme
+tx-sender
+```
+
+Importantly, the `tx-sender` variable does not change during
+inter-contract calls. This means that if you call a transaction in a
+given smart contract, that transaction is able to make calls into
+other smart contracts on your behalf. This enables a wide variety of
+applications, but it comes with some dangers for users of smart
+contracts. However, as mentioned before, the static analysis
+guarantees of our smart contracting language allow clients to know a
+priori which transactions a given smart contract will ever call. Good
+clients should always warn users about any potential side effects
+of a given transaction.
+
+Note that most assets in the smart contracting language and blockchain
```

Comment:
Is there an example of an asset that is not owned?

Reply:
This is debatable.  Some namespaces can accrue funds from name registration fees, and namespaces are definitely created by spending Stacks.  They are also "locked" when they are created, so that only the creator can import names.

Reply:
Oh -- actually, that line is:

```
Note that most assets in the smart contracting language and blockchain
will be "owned" by objects of the principal type, meaning that any
object of the principal type may own an asset.
```

Means that if someone defines something like a crypto-widget and have a table of widgets mapped to owner principals, then other smart-contracts get the ability to 'own' those widgets for free.

However, you can think of cases where widgets are 'owned' by something else. Like cryptokitties who own hats.

Reply:
I'm thinking from this example that it's fair to say that every asset has an owner (i.e. the principal that created it), but the owner is not necessarily guaranteed to be able to do anything with it (and the owner can be a contract, or the null address).

Reply:
I don't think that sentence is really trying to say anything about whether or not assets are usually owned --- it's trying to say that _principals_ are the types that usually own things.

Reply:
Does this tweak to that text clarify this?

```
Assets in the smart contracting language and blockchain will be
"owned" by objects of the principal type. Any object of
the principal type may own an asset. For the case of public-key hash
and multi-signature Stacks addresses, a given principal can operate on
their assets by issuing a signed transaction on the blockchain. _Smart
contracts_ may also be principals (reprepresented by the smart
contract's identifier), however, there is no private key associated
with the smart contract, and it cannot broadcast a signed transaction
on the blockchain.
```